presto-jdbc/src/main/java/com/facebook/presto/jdbc/WarningsManager.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.jdbc;
+
+import com.facebook.presto.spi.PrestoWarning;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.sql.SQLWarning;
+import java.util.List;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class WarningsManager
+{
+    private WarningManager connectionWarningManager = new WarningManager();
+    private WarningManager statementWarningManager = new WarningManager();
+    private WarningManager resultSetWarningManager = new WarningManager();
+    private int warningsSeen;
+
+    public void addWarnings(List<PrestoWarning> warnings)
+    {
+        for (; warningsSeen < warnings.size(); warningsSeen++) {
+            switch (warnings.get(warningsSeen).getWarningCode().getWarningType()) {
+                case CONNECTION:
```

Comment:
This is not the right model. The connection vs statement vs resultset classification is not with respect of the type of warning (the Presto engine should have no specific knowledge about these anyway), but on what action caused the warning to be emitted. From the javadocs:

* ResultSet.getWarnings()

> Retrieves the first warning reported by calls on this ResultSet object.
> [...]
> Note: This warning chain only covers warnings caused by ResultSet methods. Any warning caused by Statement methods (such as reading OUT parameters) will be chained on the Statement object.

* Statement.getWarnings()

> Retrieves the first warning reported by calls on this Statement object.
> [...]
> Note: If you are processing a ResultSet object, any warnings associated with reads on that ResultSet object will be chained on it rather than on the Statement object that produced it.

* Connection.getWarnings()

> Retrieves the first warning reported by calls on this Connection object.

Reply:
Updated. Due to #11533 warnings are only accessible via the ResultSet object.